File: medium/1292-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold/solution.py
from typing import List


class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        """
        1292. Maximum Side Length of a Square with Sum Less than or Equal to Threshold
        Time: O(m × n × log(min(m, n)))
        Space: O(m × n)
        """
        m, n = len(mat), len(mat[0])

        pref = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                pref[i][j] = (
                    mat[i - 1][j - 1]
                    + pref[i - 1][j]
                    + pref[i][j - 1]
                    - pref[i - 1][j - 1]
                )

        def possible(L: int) -> bool:
            for i in range(m - L + 1):
                for j in range(n - L + 1):
                    total = (
                        pref[i + L][j + L]
                        - pref[i][j + L]
                        - pref[i + L][j]
                        + pref[i][j]
                    )
                    if total <= threshold:
                        return True
            return False

        low, high = 0, min(m, n)
        ans = 0

        while low <= high:
            mid = (low + high) // 2
            if possible(mid):
                ans = mid
                low = mid + 1
            else:
                high = mid - 1

        return ans

File: medium/1292-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold/solution.java
/**
 * 1292. Maximum Side Length of a Square with Sum Less than or Equal to Threshold
 * Time: O(m × n × log(min(m, n)))
 * Space: O(m × n)
 */
class Solution {
    private int[][] pref;
    private int m, n, threshold;

    public int maxSideLength(int[][] mat, int threshold) {
        this.m = mat.length;
        this.n = mat[0].length;
        this.threshold = threshold;

        pref = new int[m + 1][n + 1];
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                pref[i][j] = mat[i - 1][j - 1]
                           + pref[i - 1][j]
                           + pref[i][j - 1]
                           - pref[i - 1][j - 1];
            }
        }

        int low = 0, high = Math.min(m, n);
        int ans = 0;

        while (low <= high) {
            int mid = (low + high) / 2;
            if (possible(mid)) {
                ans = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return ans;
    }

    private boolean possible(int L) {
        for (int i = 0; i <= m - L; i++) {
            for (int j = 0; j <= n - L; j++) {
                int total = pref[i + L][j + L]
                          - pref[i][j + L]
                          - pref[i + L][j]
                          + pref[i][j];
                if (total <= threshold) {
                    return true;
                }
            }
        }
        return false;
    }
}
